File: Java/198.打家劫舍.java
import java.lang.reflect.Array;
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=198 lang=java
 *
 * [198] 打家劫舍
 */

// @lc code=start
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0 || nums == null)
            return 0;
        if (nums.length == 1)
            return nums[0];
        if (nums.length == 2)
            return Math.max(nums[0], nums[1]);
        int[] sum = new int[nums.length + 1];
        if (nums.length == 3)
        return Math.max(Math.max(sum[0], sum[1]), sum[2]);
        sum[0] = nums[0];
        sum[1] = nums[1];
        sum[2] = nums[2] + nums[0];
       
        int len = nums.length;
        for (int i = 3; i < len; i++) {
                sum[i] = Math.max(sum[i - 3], sum[i - 2]) + nums[i];
            }

        return sum[len-2] > sum[len - 1] ? sum[len-2] : sum[len - 1];

    }
}
// @lc code=end

File: Cpp/198.打家劫舍.cpp
/**
	动态规划
**/

class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0) return 0;

        int dp[nums.size() + 2];
        memset(dp, 0, sizeof(dp));

        dp[1] = nums[0];
        for (int i = 2;i <= nums.size();i++) {
            dp[i] = max(dp[i - 1],nums[i - 1] + dp[i - 2]);
        }
        return dp[nums.size()];
    }
};
